### ISSUE TITLE
Any script which could help debug?

### ISSUE THREAD

**Issue by FraGard** [OSeMOSYS/OSeMOSYS_GNU_MathProg#34]
In the Energy Systems Analysis courses at KTH we have students create simple energy / CLEWs models with OSeMOSYS. They have often bugs and we are seeing that some steps when debugging seem to recur:
- If the model does not run, first check the naming;
- If it still does not run, check the Backstops are properly set at all 'energy levels';
- If it still does not run, often there are conflicts between min / max capacity or activity limits;
- If the above is ok, the model should in most cases run. But maybe some technology does not come in when it should, or viceversa, or it comes in in a strange way. For that, I'd check that input/output activity ratios are set so that the resulting RES is what we expect.

I imagine there could be tools and scripts to help in these steps?

**Comment by willu47** [comment]
Hi @FraGard - I think we could definitely write some validation type scripts to help with this. Perhaps we should create separate feature issues for each of these over at osemosys/otoole to develop further?

Otoole already has a naming convention check, which I can demonstrate. Otherwise, a script to check parameter values should be pretty trivial to write.

**Comment by willu47** [comment]
@vignesh1987 - here is the issue related to your suggestion in the steering committee meeting today.  My suggestion was to look into the built-in parameter statement definition in the GMLP.pdf document supplied with GNU MathProg.

The examples given show some ways of adding relation expressions:

```ampl
param units{raw, prd} >= 0;
param N := 20 integer >= 0 <= 100;
param p{i in I, j in J}, integer, >= 0, <= i+j, in A[i] symdiff B[j], in C[i,j], default 0.5 * (i + j);
param month symbolic default ’May’ in {’Mar’, ’Apr’, ’May’};
```
which specifies a condition that restricts the parameter or its members to satisfy that condition;

**Comment by willu47** [comment]
There's also a `check` statement.

```ampl
check: x + y <= 1 and x >= 0 and y >= 0;
check sum{i in ORIG} supply[i] = sum{j in DEST} demand[j]; 
check{i in I, j in 1..10}: S[i,j] in U[i] union V[j];
```

The check statement allows checking the resultant value of an logical expression specified in the statement. If the value is false, an error is reported.

If the subscript domain is not specified, the check is performed only once. Specifying the subscript domain allows performing multiple check for every n-tuple in the domain set. In the latter case the logical expression may include dummy indices introduced in corresponding indexing expression.

**Comment by vignesh1987** [comment]
Thanks for the pointers. I will write down the checks statements for the most common parameter inputs.

**Comment by vignesh1987** [comment]
I am creating the check statements for debugging. The check statements seem to work for all those technologies that have been explicitly declared. i.e If a parameter (let's say total annual max capacity) is not defined for a technology (hence it is supposed to take the default value), the check statement throws an error for that technology. I am sure there is a definitely a simple workaround. Does anyone know how to make the check statement consider the default values for techs whose parameters have not been explicitly declared?

**Comment by willu47** [comment]
Fixed by #50 and #58
